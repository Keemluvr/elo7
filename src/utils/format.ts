/**
 * Capitalizes the first letter of the input string and converts the rest to lowercase.
 *
 * @param {string} string - The input string to be capitalized.
 * @return {string} The capitalized string.
 */
export const capitalize = (string: string) =>
  string.charAt(0).toUpperCase() + string.slice(1).toLowerCase();

/**
 * Capitalizes each word in a string, including words separated by hyphens.
 *
 * @param {string} string - The input string to be capitalized.
 * @return {string} The capitalized string with each word capitalized and words separated by hyphens preserved.
 */
export const capitalizeWords = (string: string) => {
  return string
    .split(" ")
    .map((word) => {
      if (word.includes("-")) {
        const parts = word.split("-");
        return parts.map((part) => capitalize(part)).join("-");
      }
      return capitalize(word);
    })
    .join(" ");
};

/**
 * Formats a location string into a city, state, and country.
 *
 * @param {string} location - The location string to format.
 * @return {string} The formatted location string.
 */
export const formatLocation = (location: string) => {
  const [city, state, country] = location.split(", ");
  if (!city || !country) return "";
  return `${capitalizeWords(city)}, ${capitalizeWords(country)}`;
};

/**
 * Replaces specific substrings in a given string with their corresponding uppercase versions.
 *
 * @param {string} string - The input string to be processed.
 * @return {string} The modified string with the special cases replaced.
 */
export const specialCases = (string: string) => {
  return string
    .replace(/fp&a/gi, "FP&A")
    .replace(/Android/gi, "Android")
    .replace(/Ios/gi, "iOS")
    .replace(/devops/gi, "DevOps")
    .replace(/Sre/gi, "SRE");
};
