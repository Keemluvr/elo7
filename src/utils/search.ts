interface Data<T> {
  [key: string]: T[];
}

/**
 * Searches the data for entries that match all the terms in the query.
 *
 * @param {Data<T>} data - The data to search through.
 * @param {string | null} query - The search query.
 * @param {(keyof T)[]} properties - The properties to search within.
 * @return {Data<T> | undefined} The filtered data based on the search query.
 */
export const searchByCombinedTerms = <T>(
  data: Data<T>,
  query: string | null,
  properties: (keyof T)[]
): Data<T> | undefined => {
  if (!query) return;

  const terms = query.toLowerCase().split(" ");
  const filteredData: Data<T> = {};

  for (const key in data) {
    const filteredEntries = data[key].filter((entry: T) => {
      const entryString = properties
        .map((prop) => (entry[prop] as string).toLowerCase())
        .join(" ");
      return terms.every((term) => entryString.includes(term));
    });

    if (filteredEntries.length > 0) {
      filteredData[key] = filteredEntries;
    }
  }
  return Object.keys(filteredData).length > 0 ? filteredData : undefined;
};
