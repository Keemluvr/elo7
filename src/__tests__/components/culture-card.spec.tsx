import { render } from "@testing-library/react";
import { verifyAttributes } from "../utils/verifyAttributes";
import CultureCard from "@/components/culture-card";

describe("CultureCard component", () => {
  const props = {
    img: "/images/culture/thumbs-up.webp",
    alt: "ícone de um polegar para cima",
    title: "Qualidade de vida",
    description: "Donec in vestibulum elit.",
    redirectTo: {
      label: "dignissim quis",
      href: "https://www.elo7.com.br/sobre",
    },
  };

  const propsWithoutRedirectTo = {
    img: "/images/culture/gallery.webp",
    alt: "ícone de uma imagem",
    title: "Resultados",
    description: "Aliquam finibus facilisis elit.",
  };

  describe("List item", () => {
    test("should render the component", () => {
      const { getByRole } = render(<CultureCard {...props} />);
      const listItemElement = getByRole("listitem", {
        name: "card de cultura",
      });
      expect(listItemElement).toBeInTheDocument();
    });

    test("should render the section with correct tag name", () => {
      const { getByRole } = render(<CultureCard {...props} />);
      const listItemElement = getByRole("listitem", {
        name: "card de cultura",
      });
      expect(listItemElement.tagName).toBe("LI");
    });
  });

  describe("Image", () => {
    test("should render the image", () => {
      const { getByAltText } = render(<CultureCard {...props} />);
      const imageElement = getByAltText(props.alt);
      expect(imageElement).toBeInTheDocument();
    });

    test("should render the image with correct properties", () => {
      const { getByAltText } = render(<CultureCard {...props} />);
      const imageElement = getByAltText(props.alt);
      const expectedSrc =
        "/_next/image?url=%2Fimages%2Fculture%2Fthumbs-up.webp&w=384&q=100";
      const attributes = [
        ["src", expectedSrc],
        ["alt", props.alt],
        ["loading", "lazy"],
        ["class", "culture-card-icon"],
      ];
      verifyAttributes(imageElement, attributes);
      expect(imageElement).toHaveAttribute("srcset");
      expect(imageElement.getAttribute("srcset")).not.toBe("");
    });
  });

  describe("Title", () => {
    test("should render the title", () => {
      const { getByText } = render(<CultureCard {...props} />);
      const titleElement = getByText(props.title);
      expect(titleElement).toBeInTheDocument();
    });

    test("should render the title with correct tag name", () => {
      const { getByText } = render(<CultureCard {...props} />);
      const titleElement = getByText(props.title);
      expect(titleElement.tagName).toBe("H2");
    });
  });

  describe("Description", () => {
    test("should render the description", () => {
      const { getByText } = render(<CultureCard {...props} />);
      const descriptionElement = getByText(props.description);
      expect(descriptionElement).toBeInTheDocument();
    });
  });

  describe("Link", () => {
    test("should render the button", () => {
      const { getByRole } = render(<CultureCard {...props} />);
      const buttonElement = getByRole("button", {
        name: props.redirectTo?.label,
      });
      expect(buttonElement).toBeInTheDocument();
    });

    test("should render the button with correct tag name", () => {
      const { getByRole } = render(<CultureCard {...props} />);
      const buttonElement = getByRole("button", {
        name: props.redirectTo?.label,
      });
      expect(buttonElement.tagName).toBe("A");
    });

    test("should render the button with correct properties", () => {
      const { getByRole } = render(<CultureCard {...props} />);
      const buttonElement = getByRole("button", {
        name: props.redirectTo?.label,
      });
      const attributes = [
        ["class", "culture-card-button"],
        ["href", props.redirectTo.href],
        ["rel", "noopener noreferrer"],
        ["role", "button"],
        ["target", "_blank"],
      ];
      verifyAttributes(buttonElement, attributes);
    });

    test("should hidden the button when the redirectTo is undefined", () => {
      const { getByRole } = render(<CultureCard {...propsWithoutRedirectTo} />);
      const itemListElement = getByRole("listitem");
      expect(itemListElement).not.toHaveRole("button");
    });
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <>
        <CultureCard {...props} />
        <CultureCard {...propsWithoutRedirectTo} />
      </>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
