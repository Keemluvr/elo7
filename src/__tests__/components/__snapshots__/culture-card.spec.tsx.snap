// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CultureCard component matches snapshot 1`] = `
<DocumentFragment>
  <li
    aria-label="card de cultura"
    class="culture-card-wrapper"
  >
    <img
      alt="ícone de um polegar para cima"
      class="culture-card-icon"
      data-nimg="1"
      decoding="async"
      height="136"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fculture%2Fthumbs-up.webp&w=384&q=100"
      srcset="/_next/image?url=%2Fimages%2Fculture%2Fthumbs-up.webp&w=256&q=100 1x, /_next/image?url=%2Fimages%2Fculture%2Fthumbs-up.webp&w=384&q=100 2x"
      style="color: transparent;"
      width="136"
    />
    <h2
      class="culture-card-title"
    >
      Qualidade de vida
    </h2>
    <p
      class="culture-card-description"
    >
      Donec in vestibulum elit.
    </p>
    <a
      class="culture-card-button"
      href="https://www.elo7.com.br/sobre"
      rel="noopener noreferrer"
      role="button"
      target="_blank"
    >
      dignissim quis
    </a>
  </li>
  <li
    aria-label="card de cultura"
    class="culture-card-wrapper"
  >
    <img
      alt="ícone de uma imagem"
      class="culture-card-icon"
      data-nimg="1"
      decoding="async"
      height="136"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fculture%2Fgallery.webp&w=384&q=100"
      srcset="/_next/image?url=%2Fimages%2Fculture%2Fgallery.webp&w=256&q=100 1x, /_next/image?url=%2Fimages%2Fculture%2Fgallery.webp&w=384&q=100 2x"
      style="color: transparent;"
      width="136"
    />
    <h2
      class="culture-card-title"
    >
      Resultados
    </h2>
    <p
      class="culture-card-description"
    >
      Aliquam finibus facilisis elit.
    </p>
  </li>
</DocumentFragment>
`;
